### DISCUSSION TITLE
New feature to propagate proxies to yarn is causing us problems

### DISCUSSION THREAD

**Discussion by lukeapage** [renovatebot/renovate#27983] [Request Help] [answered]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

 37.232.0

### Please tell us more about your question or problem

This Pr is breaking our setup: https://github.com/renovatebot/renovate/pull/27794#issuecomment-1999524412

We have an internally hosted npm repo but requirement for anything external to go via a proxy. So usually we don't have a proxy configured in yarn, but do have one configured on the environment variable.

As stated, yarn doesn't have no-proxy support - but it does have this option:
https://github.com/yarnpkg/berry/issues/1530#issuecomment-1415692471

whilst testing we also found an issue because renovate would edit the home directory of the build agent and then it doesn't undo its changes (In fact, we found the config file grows by 1 line with every renovate run...)

I see a few ways this could be fixed:

1. We could add the above configuration to all our repositories to undo the change by renovate (the constant growing config file would still be an issue)
2. An option could be added to renovate not to copy the proxy into yarn so we can disable the behaviour
3. Renovate could look at NO_PROXY and also propagate this config into the yarn config as in the issue link

**Comment by rarkins** [comment]
I think that propagating the NO_PROXY value makes sense regardless.

Renovate could/should also undo its changes to the global config although that's a little challenging as right now it does the settings through running `yarn` commands instead of finding and changing the file directly. I'm not sure if it could undo the changes through `yarn` commands too?

These changes can be made separately (and you are welcome to contribute a PR if it's urgent to you).

As a short term workaround we could add an experimental flag for self-hosters which skips this new capability

**Reply by rarkins** [reply]
BTW I think it's unfortunate behavior of Yarn if running the same command multiple times _appends_ to the config file instead of updating it.

**Comment by lukeapage** [comment]
Thanks - its not urgent because we can just not upgrade renovate for a while.

If you think you will do something that will fix this problem in the next month I'll leave it to you, otherwise I can schedule some work to contribute a PR.

**Reply by rarkins** [reply, marked as answer]
https://github.com/renovatebot/renovate/pull/27984

**Comment by fgreinacher** [comment]
Is https://github.com/renovatebot/renovate/pull/27984 considered a short-term fix?

I don't think Renovate should evaluate HTTP_PROXY without also evaluating NO_PROXY. I strongly assume most environments that require configuring a proxy also need exceptions. So all of those will be broken now.

I would prefer to have this as an opt-in rather than an opt-out. I'd even consider this a breaking change TBH.

**Reply by viceice** [reply]
@rarkins sounds good to me to make it opt-in

**Reply by rarkins** [reply]
I agree, will change it

**Reply by rarkins** [reply]
https://github.com/renovatebot/renovate/pull/28100
